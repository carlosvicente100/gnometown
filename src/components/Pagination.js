import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

const PaginationContainer = styled.div`
  background-color: #2e8acc;
  display: flex;
  justify-content: space-between;
  width: 100%;
`
const PaginationList = styled.ul`
  color: white;
  display: flex;
  justify-content: space-between;
  list-style-type: none;
  overflow: hidden;
  li {
    cursor: pointer;
    padding: 5px;
  }
  li:hover {
    background-color: #2980b9;
  }
  li.active {
    background-color: #2980b9;
  }
`
const PaginationButton = styled.button`
  background-color: #2980b9;
  border: none;
  color: white;
  padding: 12px 16px;
  font-size: 16px;
  cursor: pointer;
  &:disabled {
    color: black;
    background-color: #7f8c8d;
    cursor: not-allowed;
  }
`

const Pagination = ({ setCurrentPage, totalPagination, current }) => {
  const changePage = (event) => {
    if (event.target.value !== current) {
      setCurrentPage(event.target.value)
    }
  }
  const decreasePage = () => {
    setCurrentPage(current - 1)
  }
  const increasePage = () => {
    setCurrentPage(current + 1)
  }

  const makeList = totalPagination.map((number, index) => (
    <li key={index} className={`${current === index ? 'active' : ''} `} onClick={changePage} value={index}>
      {index + 1}
    </li>
  ))

  return (
    <PaginationContainer className="Pagination">
      <PaginationButton disabled={current !== 0 ? false : true} onClick={decreasePage}>{`<`}</PaginationButton>
      <PaginationList>{makeList}</PaginationList>
      <PaginationButton
        disabled={current !== totalPagination.length - 1 ? false : true}
        onClick={increasePage}
      >{`>`}</PaginationButton>
    </PaginationContainer>
  )
}
Pagination.propTypes = {
  setCurrentPage: PropTypes.func.isRequired,
  totalPagination: PropTypes.array.isRequired,
  current: PropTypes.number.isRequired
}

Pagination.defaultProps = {}

export default Pagination
