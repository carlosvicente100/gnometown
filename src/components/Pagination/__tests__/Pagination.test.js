import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Pagination from '../Pagination'
import fixtureItemList from '../../../fixtures/gnomes'

const dummyProps = {
  setCurrentPage: jest.fn(),
  totalPagination: [[fixtureItemList], [fixtureItemList]],
  current: 1 //second position
}

beforeEach(() => {
  jest.clearAllMocks()
})

describe('Pagination Component Suite', () => {
  test('Snapshot check', () => {
    const component = renderer.create(<Pagination {...dummyProps} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('Pagination Structure', () => {
    test('Should contain some elements - two buttons and a list of pages with current active ', () => {
      const { getByText } = render(<Pagination {...dummyProps} />)
      expect(getByText('<')).toBeInTheDocument()
      expect(getByText('>')).toBeInTheDocument()
      expect(getByText('1')).toBeInTheDocument()
      expect(getByText('2')).toBeInTheDocument()
      expect(getByText('2')).toHaveClass('active') //second position
    })
  })
  describe('Pagination Functionallity', () => {
    test("Shouldn't call setCurrentPage on click next Page from Last Page", () => {
      const { getByText } = render(<Pagination {...dummyProps} />)

      expect(getByText('>')).toHaveAttribute('disabled')
      fireEvent.click(getByText('>'))

      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(0)
    })

    test('Should call setCurrentPage on click prev Page from Last Page', () => {
      const { getByText } = render(<Pagination {...dummyProps} />)

      expect(getByText('<')).not.toHaveAttribute('disabled')
      fireEvent.click(getByText('<'))

      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(1)
      expect(dummyProps.setCurrentPage).toHaveBeenCalledWith(dummyProps.current - 1)
    })

    test('Should call setCurrentPage on click different page value from current', () => {
      const { getByText } = render(<Pagination {...dummyProps} />)
      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(0)

      expect(getByText('1')).not.toHaveClass('active')
      fireEvent.click(getByText('1'))
      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(1)
      expect(dummyProps.setCurrentPage).toHaveBeenCalledWith(dummyProps.current - 1)
    })

    test("Shouldn't call setCurrentPage on click same page value from current", () => {
      const { getByText } = render(<Pagination {...dummyProps} />)
      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(0)

      expect(getByText('2')).toHaveClass('active')
      fireEvent.click(getByText('2'))
      expect(dummyProps.setCurrentPage).toHaveBeenCalledTimes(0)
    })
  })
})
