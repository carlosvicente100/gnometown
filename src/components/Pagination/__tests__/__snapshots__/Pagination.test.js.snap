// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component Suite Snapshot check 1`] = `
.c0 {
  background-color: #2e8acc;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c2 {
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  list-style-type: none;
  overflow: hidden;
}

.c2 li {
  cursor: pointer;
  padding: 5px;
}

.c2 li:hover {
  background-color: #2980b9;
}

.c2 li.active {
  background-color: #2980b9;
}

.c1 {
  background-color: #2980b9;
  border: none;
  color: white;
  padding: 12px 16px;
  font-size: 16px;
  cursor: pointer;
}

.c1:disabled {
  color: black;
  background-color: #7f8c8d;
  cursor: not-allowed;
}

<div
  className="c0 Pagination"
>
  <button
    className="c1"
    disabled={false}
    onClick={[Function]}
  >
    &lt;
  </button>
  <ul
    className="c2"
  >
    <li
      className=" "
      onClick={[Function]}
      value={0}
    >
      1
    </li>
    <li
      className="active "
      onClick={[Function]}
      value={1}
    >
      2
    </li>
  </ul>
  <button
    className="c1"
    disabled={true}
    onClick={[Function]}
  >
    &gt;
  </button>
</div>
`;
